### DISCUSSION TITLE
18.1

### DISCUSSION THREAD

**Discussion by carloscabanero** [blinksh/blink#2067] [Announcements]
# New FileProvider Integration

We're excited to introduce the new File Provider Replicated extension! We are taking this opportunity to bring more improvements, streamlining file management and resolving many longstanding issues with FileProviders in Blink. Now is the perfect time to report any additional issues, as we’re actively working to refine the experience. Here’s a look at some of the key use cases improved with the API:

- **More stable API**: The new API is more robust by separating responsibilities between the System and the Extension.
- **Improvements and Fixes**: The new Replicated File Provider implements many previously suggested ideas, such as better progress control for file operations.
- **Automatic sync between local and remote**: Changes made on either the remote or local side will sync automatically without needing to manually navigate the filesystem. A pull operation is performed every 5 seconds to keep files up-to-date.
- **Automatic re-downloads**: For example, if you modify and recompile LaTeX PDFs or image files on the remote, they will automatically re-download and update in your preview, ensuring you always have the latest version at hand.
- **Kernel-level integration**: Coming next. The new File System is a user-level mount, meaning that common POSIX operations like “fopen” work transparently with the remote file system. We will leverage this in Blink to use our CLI tools, including Vim, to browse and edit the remote file system right from the shell.

## How to Enable

1. Go to **Config > Host**.
2. Scroll to the bottom to find the **Files.app extension**.
3. Tap into the Domain settings and enable the **"Replicated"** switch.
    - If authentication fails, check that you can access the server directly and that SFTP access is enabled.

*Note: This option will be enabled by default in future versions.*

### How to Reset

If you want to reset the integration, remove the domain by swiping it left. This will clear the current state, including any downloaded files, and give you a fresh start. Be aware that pending local files may be lost, so consider moving them elsewhere first.

## Testing Guidelines

- The integration, even if in Beta, is stable for daily use. We’re aware of a few corner cases and have plans to improve performance, but these shouldn’t impact regular use. Since the implementation is stateless, it should recover well if something goes wrong.
    - Use the **Files.app** to access remote files. Common operations like Copy, Move, Delete, and Rename should work smoothly.
    - You can control the progress of some operations, like stopping downloads and uploads.
    - Help us ensure compatibility with other apps for file downloads, modifications, and new file creation.

- **Temporary files**: The integration creates temporary files for uploads at the root of the file system, named in the format “.blink.tmp.xxxx.” These can be safely deleted from time to time. Future updates will attempt to clean up these files automatically when operations fail, but it’s still good to know where they are.
  
- **Storage Management**: The operating system now manages file storage, meaning downloaded files’ lifetimes are automatically handled. Deleting the integration will free up space. In upcoming versions, we’ll offer additional options to manage storage and evict files as needed.

## Known Issues

- **Renaming Folders**: Folder renames work, but their contents may need to be re-downloaded. This occurs because our WorkingSet may pull data before the operation commits fully.
- **Symlink Support**: Currently in development.
- **Performance**: While the new provider can handle multiple operations at once, larger structures or multiple simultaneous copies may cause delays as operations complete in blocks. We’re exploring ways to improve this.
- **Efficient Transfers**: For high-performance file transfers, especially for large or multiple files, SCP or SFTP remain the best options.

## Reporting Errors

If a file or operation fails to complete:
- Restart the provider and try again.
- Test on a different server or in a different location.
- Open an issue and, if possible, include the relevant files or detailed steps to help us replicate the issue. 

Thanks for your feedback!

**Comment by mistakenideas** [comment]
Great, look forward to giving it a whirl. A quick one, when navigating to a previously configured (old) Blink host share in Files.app, the popup has a formatting issue (step 3) and shows an error message, not sure if this is intentional?
![IMG_3474](https://github.com/user-attachments/assets/b101e75d-5b65-4f3d-8d8e-7a68aabcbcd3)

I’ve also tried adding a new shared folder but keep getting this message when navigating to it in Files.app: Content Unavailable - The folder contents could not be displayed because of an unknown error. 
I removed the old share from Files.app and also removed the new one from Blink and recreated it but no luck so far. 
I’m still on iPadOS 17.7 if that makes any difference?

**Reply by carloscabanero** [reply]
This is a different topic from the current release, please create a separate request for it. In your case, it is usually a problem with SFTP. Please make sure you can use sftp in the shell, etc... https://docs.blink.sh/advanced/files-app

**Comment by carloscabanero** [comment]
After installing the new build, please to remove the previous Files.app location so the new updates (specially "ghost folders"), are applied correctly.

Build 949

- **Fixed ghost folders**: Resolved an issue where numbered "ghost" folders (e.g., "Folder name 2") would appear at the root when first starting the extension.

- **Improved folder renaming**: Previously, renaming a folder could result in contents needing to be re-downloaded if background updates occurred simultaneously. Now, contents are reliably retained after renames, ensuring 100% functionality for this feature.

**Comment by carloscabanero** [comment]
Thanks to everyone on both TestFlight and Discord leaving feedback that is helping us take the final steps towards the release. In this release all of the internals should be ready to go. The next release should be the final one, and will focus on improving the final experience.

Build 957

- **Veracrypt support**: Revised compatibility with Veracrypt. For more details, please refer to #2033).

- **Symlink browsing**: You can now browse symbolic links. Note that other operations, such as renaming, reparenting, or deleting symlinks, will not be supported.

- **Automatic file change detection**: The system now automatically fetches changes for individual, previously downloaded files.

- **Folder downloads**: Resolved an issue where downloading folders with complex structures would get stuck. Please note that due to a system bug, progress updates may not reflect correctly, and operations could be automatically canceled after 120 seconds. Retrying the operation will succeed.

- **Empty folder updates**: Fixed an issue where empty folders were not automatically updating to show new files created within them.

- **SSH connection cleanup**: SSH connections after the FP terminates will now properly clean up in all scenarios.

- **Improved error handling**: Enhanced error reporting and communication to users.

- **Random crash fixes**: Addressed crashes caused by database issues.

- **Performance improvements for create/modify operations**: The system now blocks conflicting concurrent operations on the same items to ensure smoother handling of file creation and modifications.

- **Optimized folder refreshes**: Improved performance during continuous refreshes, particularly when multiple folders are open simultaneously.

## Known Issues
- **Old file provider**: The old file provider cannot coexist with this one. Please, remove the previous implementations or trigger the switch to enable the Replicated version before browsing.
- **Efficient Transfers**: For high-performance file transfers, especially for large or multiple files, SCP or SFTP remain the best options.

**Comment by carloscabanero** [comment]
## Build 959

- **New FileProvider as default**: The new FileProvider is now the only option, replacing the old provider. Migration happens automatically and should "Just Work".

- **SFTP connection validation**: You can now validate SFTP connections directly within the UI by navigating to **Host > Files.app configuration**. This will make it easier to find connection issues without going to the logs.

- **Logs per domain**: Logs are now available per domain too, and can be found in the `.blink` folder, named `fp-<location_name>-<location_id>.log`.

- **Jump host support**: The new FileProvider now includes support for jump hosts.

- **Shortcut integration**: Added support for the "Get Contents of Folder" action in Shortcuts. (Issue #2065)

- **Improved ProxyCommand and ProxyJump**: Enhanced handling for `ProxyCommand` and `ProxyJump` on SCP/SFTP in the CLI. (Issue #2075).

- **Symbolic link deletion**: Symbolic links can now be deleted through the provider. This is specially important when deleting large structures containing symbolic links, that would get blocked before.

- **ProxyJump SSH connection handling**: Fixed an issue where SSH would not properly close the connection when using ProxyJump. (Issue #2034)

- **Memory fixes**: Addressed memory management issues for both SSH and SFTP.

## Notes
- **Replication**: While the new File Provider ensures seamless replication between the old and new systems, it involves significant internal operations to maintain this consistency. For large-scale file transfers or operations, tools like SCP or SFTP are still the preferred options for efficiency. Similarly, for handling large file structures, performing actions like "cp" or "rm" directly at the destination offers much better performance.

**Comment by carloscabanero** [comment]
## Build 963

### Bookmarked Locations
You can now access remote file systems directly from the Blink Shell using Bookmarks, in addition to the Files.app. Thanks to the kernel-level integration of the Replicated File Provider extension, local shell commands like `ls`, `cd`, `find`, and even `vim` work seamlessly with bookmarked locations. 
- Bookmarks can include any Blink folder from a remote system or even file locations exposed by other applications. 
- This feature replaces the previous `link-files` command behavior and was a key reason for migrating to the new Replicated File Provider Extension API. 

To add a bookmark: 1. Go to **Settings > Bookmarks** and create a new bookmark. 2. Select the location in Files.app and press **Open**. 3. Give the symlink a name. 4. Return to the shell to browse the directory as if it were any other local directory. 

We're really excited about this feature and please let us know how you use it!

### Known Issues
- The app may crash when used with certain local locations. We are investigating.
- Beware of running operations that may need to trigger downloads of each and every file, like `grep` in a large structure which hasn't been previously downloaded.

**Comment by carloscabanero** [comment]
## Build 965

This will probably be the last version before public release. We appreciate once more everyone's help with testing!

- **Custom Keyboard Presses**: Added support for `Home`, `End`, `PgUp`, and `PgDown` keys. You can now create custom key presses using these keys, along with modifiers like `Alt-Home`, `Alt-End`, etc.

- **Local File System Bookmarks Improvements**: Addressed issues with certain locations not working properly, thanks to feedback from testers.

- **Fixed Home/Documents Link**: Resolved an issue where the Documents link was not working in build 963.

- **Software Keyboard Fix in Settings**: Fixed an issue where the software keyboard would appear when returning to Blink with the Settings screen open.

- **Cmd Key Fix in Settings**: Corrected a bug where the `Cmd` key would get stuck when opening Settings with `Cmd-,`, causing unintended shortcuts. The key now properly releases when returning to the shell.

### Notes
- Beware of running operations that may need to trigger downloads of each and every file, like `grep` in a large structure which hasn't been previously downloaded.

**Comment by carloscabanero** [comment]
## Build 969

- **Temporary File Cleanup**: The provider now properly cleans up leftover temporary files (from failed operations) in the remote home folder. For reference, these files have the `.blink.tmp` prefix.

- **Fixed Local Device Bookmarks**: Resolved an issue where bookmarking local folders on the device could fail due to path handling errors.

- **File Provider Failed Operations Handling**: Irrecoverable operations are now marked as failures, preventing unnecessary retries until changes are made.

- **Bookmarks UI Improvements**: Enhanced the user interface and flows for saving changes to bookmarks. Thanks for the feedback.

**Comment by carloscabanero** [comment]
**Please note:** Build 973 was never released to testers due to an issue with the order of submission. The next release, v18.1.1 (build 977), will include these fixes and ensure that the app is aligned with the App Store once again.

## Build 973 and 974 (AppStore)

Pushed as 18.1.0 final.

- **Pragmata Pro 830**: Regression in the build scripts, we were downloading the previous version for Pragmata #2085.

- **Fixed background crash in Files.app extension**: When the Provider would be initialized in the background, the thread may crash.

**Comment by carloscabanero** [comment]
## v18.1.1 Build 977

- Fixes SFTP and SCP issues where wildcards would not work properly on source arguments.

**Comment by carloscabanero** [comment]
## v18.1.1 Build 979 - 980

- Added Bookmarks on all plans (Blink Classic).
